DEVDOCS-4979 [new]: GQL Storefront API, add queries for site content (#1973)
bigcommerce/dev-docs: docs/api-docs/storefront/graphql/graphql-site-content.mdx

# Site Content with the GraphQL Storefront API

BigCommerce's GraphQL Storefront API allows you to retrieve site content from each storefront channel. Having access to site content makes it easier to build headless storefront applications using BigCommerce's built-in content, such as theme widgets.

Developers can use frontend GraphQL to perform the following: 

> - Query [web page content]((https://support.bigcommerce.com/s/article/Web-Pages?language=en_US)), including plain-text pages, contact form pages, blog posts, and more

Reviewer: ```suggestion
- Query [web page content]((https://support.bigcommerce.com/s/article/Web-Pages?language=en_US)), including HTML pages, contact forms, blog posts, and more
```

- Retrieve [banner content](https://support.bigcommerce.com/s/article/Creating-Editing-Banners?language=en_US) on home, search, category, and brand pages 

> - Retrieve rendered [widgets](https://support.bigcommerce.com/s/article/Page-Builder?language=en_US#builder) built using Page Builder

Reviewer: ```suggestion
- Retrieve rendered [widgets](https://support.bigcommerce.com/s/article/Page-Builder?language=en_US#builder) built using Page Builder or [Widgets API](https://developer.bigcommerce.com/api-docs/store-management/widgets/widgets-overview)
```

- Use the web page URL to retrieve web pages and blog posts

For more info on web pages, see [Page Types](https://developer.bigcommerce.com/docs/rest-content/pages#page-types). Queries don't support RSS-syndicated content pages, as these are deprecated.

## Web pages

> Queries for web pages retrieve web pages for a specific storefront. The request URL contains the storefront domain. For example, to retrieve web pages from the `yourstore.example.com` storefront domain, send a request to the URL `https://{yourstore.example.com}/graphql`. If the request URL uses the permanent store URL, the query returns web pages from the default storefront channel.

Reviewer: I think this should be rewritten for clarity. Keep in mind that most headless storefronts use channel canonical URLs.

I would try to convey:
- Storefront GraphQL queries are in the context of a particular Channel, and will only return web pages related to that channel ID
- We don't need to mention the different URLs here, maybe we can just link here to help people understand which URL to use: https://developer.bigcommerce.com/api-docs/storefront/graphql/graphql-api-overview#faq-querying-for-channels-other-than-stencil-storefronts-and-querying-from-external-systems

Author: ✔️  rewritten & used link instead

<Tabs items={['Request', 'Response']}>
<Tab>

```graphql filename="Example query: Get web pages" showLineNumbers copy
query {
  site {
    content {
      pages (filters: {isVisibleInNavigation: true }) {
        edges {
          node {
            __typename
            entityId
            name
            parentEntityId
            seo {
              pageTitle
              metaKeywords
              metaDescription
            }
            ... on NormalPage {
              plainTextSummary (characterLimit: 100)
              path
            }
            ... on ExternalLinkPage {
              link
            }
            ... on ContactPage {
              path
              plainTextSummary (characterLimit: 100)
              contactFields
            }
            ... on RawHtmlPage {
              path
              plainTextSummary (characterLimit: 100 )
            }
            ... on BlogIndexPage {
              path
            }
          }
        }
      }
    } 
  }
}
```

</Tab>
<Tab>

```json filename="Example query: Get web pages" showLineNumbers copy

> # The response orders web pages by entityId, or the order in which a merchant created the web page

Reviewer: ```suggestion
# The response orders web pages ordered by entityId, with the lowest IDs returned first
```

                "contactFields": [
                  "fullname",
                  "phone"
                ]
              }
            },
            {
              "node": {
                "__typename": "ExternalLinkPage",
                "entityId": 8,
                "name": "Youtube page",
                "parentEntityId": null,
                "seo": {
                  "pageTitle": "",
                  "metaKeywords": "",
                  "metaDescription": ""
                },
                "link": "https://www.youtube.com/"
              }
            },
            {
              "node": {
                "__typename": "RawHtmlPage",
                "entityId": 9,
                "name": "Random page",
                "parentEntityId": null,
                "seo": {
                  "pageTitle": "",
                  "metaKeywords": "",
                  "metaDescription": ""
                },
                "path": "/random-page/",
                "plainTextSummary": "This is a random page.\r\n\r\n\r\n\r\n\r\nThis is the body of a random page."
              }
            }
          ]
        }
      }
    }
  }
}
```

</Tab>
</Tabs>

### Rendered regions

> The following example retrieves rendered widgets on plain-text pages.

Reviewer: ```suggestion
The following example retrieves rendered widgets on plain-text pages. This will return the rendered HTML content of any widgets which have [Placements](https://developer.bigcommerce.com/api-docs/store-management/widgets/tutorials/widgets-tutorial#create-a-placement) in a given page's Regions.
```

                    }
                  ]
                }
              }
            }
          ]
        }
      }
    }
  }
}
```

</Tab>
</Tabs>

> The following example retrieves rendered widgets on contact form pages.

Reviewer: This example might be somewhat duplicative, should the reader be able to infer this from the previous example?

Author: True, I decided to remove it ✔️

<Tab>

```json filename="Example query: Get contact form page" showLineNumbers copy
{
  "data": {
    "site": {
      "route": {
        "node": {
          "id": "Q29udGFjdFBhZ2U6Ng==",
          "entityId": 6,
          "parentEntityId": null,
          "name": "Contact Us Form",
          "isVisibleInNavigation": true,
          "seo": {
            "metaKeywords": "Contact Us",
            "metaDescription": "Contact Info"
          },
          "path": "/contact-us-form/",
          "plainTextSummary": "Visitors can use the form on this page to submit questions or feedback.&nbsp;\r\nStore is located in...",
          "contactFields": [
            "fullname",
            "phone"
          ]
        }
      }
    }
  }
}
```

</Tab>
</Tabs>

<Tabs items={['Request', 'Response']}>
<Tab>

```graphql filename="Example query: Get raw HTML page" showLineNumbers copy
query {
  site {
    route(path: "/random-page/") {
      node {
        id
        ... on WebPage {
          entityId
          parentEntityId
          name
          isVisibleInNavigation
          seo {
            metaKeywords
            metaDescription
          }
        }
        ... on RawHtmlPage {
		  path

>           plainTextSummary (characterLimit: 100)

Author: Good point, I removed plainTextSummary and added htmlBody for the raw pages 

@jmwiese Tagging you here for visibility (I think the field currently exists on the type because I see it in the Playground Schema)

        }
      }
    }
  }
}
```

</Tab>
<Tab>

```json filename="Example query: Get raw HTML page" showLineNumbers copy
{
  "data": {
    "site": {
      "route": {
        "node": {
          "id": "UmF3SHRtbFBhZ2U6OQ==",
          "entityId": 9,
          "parentEntityId": null,
          "name": "Random page",
          "isVisibleInNavigation": true,
          "seo": {
            "metaKeywords": "",
            "metaDescription": ""
          },
          "path": "/random-page/",
          "plainTextSummary": "This is a random page.\r\n\r\n\r\n\r\n\r\nThis is the body of a random page."
        }
      }
    }
  }
}
```

</Tab>
</Tabs>

## Blogs

Queries for blogs retrieve published blog posts and exclude draft blogs.

> Queries return blogs even if they are hidden from the storefront's navigation menu, as the main blog index page and blog posts are still accessible by URL.

Author: ✔️ Just tested it & edited it to say published blogs

                    "metaKeywords": "Best Reviews,Most Popular",
                    "metaDescription": "Flower products"
                  }
                }
              },
              {
                "node": {
                  "entityId": 1,
                  "name": "Indoor Planting",
                  "tags": [
                    "Holiday",
                    "Lowest Price"
                  ],
                  "htmlBody": "<p>1. Pay attention to humidity<br>2. Provide adequate lighting</p>",
                  "plainTextSummary": "1. Pay attention to humidity2. Provide adequate lighting",
                  "author": "Anonymous",
                  "path": "/blog/indoor-planting/",
                  "publishedDate": {
                    "utc": "2014-02-15T19:31:55Z"
                  },
                  "thumbnailImage": null,
                  "seo": {
                    "pageTitle": "Indoor Planting",
                    "metaKeywords": "Holiday,Lowest Price",
                    "metaDescription": ""
                  }
                }
              }
            ]
          }
        }
      }
    }
  }
}
```

</Tab>
</Tabs>

> ### Rendered regions

Reviewer: Since the Blog templates in Cornerstone don't contain any regions by default, I'm wondering if we should leave this section out. Will it just confuse people?
https://github.com/bigcommerce/cornerstone/blob/master/templates/pages/blog-post.html

Author: ✔️  Thanks for mentioning. If it's not default behavior, it might confuse people, so I took it out.

            }
          ]
        }
      }
    }
  }
}
```

</Tab>
</Tabs>

> ### Routes

Author: ✔️ Sounds like a good idea, I put a route example query that works for any of the types (only 1 example in the doc now)

                  "phone"
                ]
              }
            },
            {
              "node": {
                "__typename": "ExternalLinkPage",
                "entityId": 8,
                "name": "Youtube page",
                "parentEntityId": null,
                "seo": {
                  "pageTitle": "",
                  "metaKeywords": "",
                  "metaDescription": ""
                },
                "link": "https://www.youtube.com/"
              }
            },
            {
              "node": {
                "__typename": "RawHtmlPage",
                "entityId": 9,
                "name": "Random page",
                "parentEntityId": null,
                "seo": {
                  "pageTitle": "",
                  "metaKeywords": "",
                  "metaDescription": ""
                },
                "path": "/random-page/",
                "htmlBody": "<html>\r\n\r\n<head>\r\n<title>\r\nExample page\r\n</title>\r\n</head>\r\n\r\n<body>\r\nExample body\r\n</body>\r\n\r\n</html>"
              }
            }
          ]
        }
      }
    }
  }
}
```

</Tab>
</Tabs>

### Rendered regions

> The following example retrieves rendered widgets on plain-text pages. This returns the rendered HTML content for all widgets that are [placed in Regions](/api-docs/store-management/widgets/tutorials/widgets-tutorial#create-a-placement) on the given page.

Reviewer 2: ```suggestion
The following example retrieves rendered widgets on plain-text pages. This example returns the rendered HTML content for all widgets that are [placed in Regions](/api-docs/store-management/widgets/tutorials/widgets-tutorial#create-a-placement) on the given page.
```

                    "metaDescription": "Flower products"
                  }
                }
              },
              {
                "node": {
                  "entityId": 1,
                  "name": "Indoor Planting",
                  "tags": [
                    "Holiday",
                    "Lowest Price"
                  ],
                  "htmlBody": "<p>1. Pay attention to humidity<br>2. Provide adequate lighting</p>",
                  "plainTextSummary": "1. Pay attention to humidity2. Provide adequate lighting",
                  "author": "Anonymous",
                  "path": "/blog/indoor-planting/",
                  "publishedDate": {
                    "utc": "2014-02-15T19:31:55Z"
                  },
                  "thumbnailImage": null,
                  "seo": {
                    "pageTitle": "Indoor Planting",
                    "metaKeywords": "Holiday,Lowest Price",
                    "metaDescription": ""
                  }
                }
              }
            ]
          }
        }
      }
    }
  }
}
```

</Tab>
</Tabs>

## Routes

Retrieve web pages, the main blog index page, and blog posts using the web page URL.

> The following query retrieves info for the page with the given path URL. 

Reviewer 2: ```suggestion
The following query retrieves information for the page with the given path URL. 
```
